### PR TITLE
[ts] Allow optional binding pattern parameters within types/interfaces

### LINKED ISSUE #12214: Optional binding pattern parameter inside TypeScript interface causes invalid parsing error
## Bug Report

**Current behavior**

When optional binding pattern parameter appears inside a TypeScript interface declaration, the following invalid parsing error is thrown.

> A binding pattern parameter cannot be optional in an implementation signature.

**Input Code**

```ts
interface Foo {
  foo([]?) 
}
```

**Expected behavior**
Parsing should succeed without any parsing error.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
{
  "presets": ["@babel/preset-typescript"]
}
```

**Environment**
```
  System:
    OS: Linux 4.4 Ubuntu 16.04.7 LTS (Xenial Xerus)
  Binaries:
    Node: 10.15.1 - ~/local/lib/node-v10.15.1-linux-x64/bin/node
    Yarn: 1.22.4 - ~/local/lib/node-v10.15.1-linux-x64/bin/yarn
    npm: 6.9.0 - ~/local/lib/node-v10.15.1-linux-x64/bin/npm
  npmPackages:
    @babel/cli: ^7.12.1 => 7.12.1 
    @babel/core: ^7.12.3 => 7.12.3 
    @babel/polyfill: ^7.12.1 => 7.12.1 
    @babel/preset-env: ^7.12.1 => 7.12.1 
    @babel/preset-typescript: ^7.12.1 => 7.12.1 
```

**Additional context**
This issue seems to be related to #11725.

### PR DESCRIPTION
**Pull request by existentialism**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12214 
| Patch: Bug Fix?          | Y
| License                  | MIT

### CODE REVIEW

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -2573,7 +2573,11 @@ export default (superClass: Class<Parser>): Class<Parser> =>
     // Allow type annotations inside of a parameter list.
     parseAssignableListItemTypes(param: N.Pattern) {
       if (this.eat(tt.question)) {
-        if (param.type !== "Identifier" && !this.state.isDeclareContext) {
+        if (
+          param.type !== "Identifier" &&
+          !this.state.isDeclareContext &&
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:2578]: Does `this.state.isDeclareContext` imply `this.state.inType`? If so we can remove `!this.state.isDeclareContext` check.

Comment (existentialism) [reply packages/babel-parser/src/plugins/typescript/index.js:2578]: Let me dig deeper, there was some failures when I replaced it originally.

Comment (JLHwung) [reply packages/babel-parser/src/plugins/typescript/index.js:2578]: Ah I see. `this.state.inType` is technically `this.state.inTypeAnnotation`, so `this.state.isDeclareContext` does not imply `this.state.inType`.

Comment (existentialism) [reply packages/babel-parser/src/plugins/typescript/index.js:2578]: Yeah, naming is hard x2 😸

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]

**Pull request merged**